iree/compiler/Dialect/Flow/Transforms/IdentifyDispatchRegions2.cpp

```diff
@@ -0,0 +1,374 @@
+  DispatchRegionOp op;
+  Operation *anchorOp;
+
+  Block &getEntryBlock() { return op.body().front(); }
+
+  static llvm::Optional<DispatchRegion> form(Operation *anchorOp) {
+    auto loc = anchorOp->getLoc();
+    if (anchorOp->getNumResults() < 1) {
+      emitError(loc) << "dispatch anchor op must have at least one result: "
+                     << *anchorOp;
+      return llvm::None;
+    }
+    Value result = anchorOp->getResult(0);
+    Value workload = calculateWorkload(anchorOp, result);
+    if (!workload) return llvm::None;
+
+    OpBuilder builder(anchorOp->getContext());
+    auto created =
+        DispatchRegionOp::formFromAnchorOp(workload, anchorOp, builder);
+    if (!created) return llvm::None;
+    return DispatchRegion{created->first, created->second};
+  }
+
+  // After a call to inlineDispatchOp, adds the results of the inlined op to
+  // the dispatch region's results and redirects any uses outside of the
+  // dispatch region.
+  void returnAndReplaceUses(Operation *origOp, Operation *inlinedOp) {
+    // Extend the arity of the dispatch region.
+    OpBuilder builder(op.getContext());
+    llvm::SmallVector<Value, 4> addlResults(inlinedOp->getResults());
+    origOp->replaceAllUsesWith(
+        DispatchRegionOp::appendResults(op, addlResults, builder));
+  }
+};
+
+// Clones and hoists any identity metadata ops from the operands and results
+// of the dispatch region back out into the surrounding block.
+// This function is not general purpose: it only knows how to undo sinking
+// done by dispatch region formation.
+void hoistDispatchRegionMetadataOps(DispatchRegion &dr,
+                                    OpDispatchPolicy &policy) {
+  BlockAndValueMapping mapping;
+  Block &block = dr.getEntryBlock();
+  for (unsigned i = 0, e = block.getNumArguments(); i < e; ++i) {
+    mapping.map(block.getArgument(i), dr.op.args()[i]);
+  }
+
+  // Hoist metadata ops from the operand edge.
+  for (auto it : llvm::enumerate(block.getArguments())) {
+    auto &blockArg = it.value();
+    for (auto &blockUse : blockArg.getUses()) {
+      Operation *useOp = blockUse.getOwner();
+      if (!policy.isIdentityMetadata(useOp) || useOp->getOperand(0) != blockArg)
+        continue;
+      OpBuilder builder(dr.op);
+      Operation *newOp = builder.clone(*useOp, mapping);
+      dr.op.argsMutable().slice(it.index(), 1).assign(newOp->getResult(0));
+    }
+  }
+
+  // Hoist metadata ops from the result edge.
+  // Since initial formation can only have a single block, this is safe.
+  auto *terminator = block.getTerminator();
+  for (auto it : llvm::enumerate(terminator->getOperands())) {
+    Operation *defOp = it.value().getDefiningOp();
+    if (!defOp || !policy.isIdentityMetadata(defOp)) continue;
+    OpBuilder builder(dr.op.getContext());
+    builder.setInsertionPointAfter(dr.op);
+    Operation *newOp = builder.clone(*defOp, mapping);
+    dr.op.getResult(it.index()).replaceAllUsesWith(newOp->getResult(0));
```

Comment:
worth having a helper for this? (in case we have a metadata op that have multiple results/operands, or where operand 0 isn't special)

Reply:
Maybe. I'm still hoping that we find a better represntation for this and will punt on a more generic helper for now (until the next time I have to do this).